lib/charms/mongodb/v0/mongodb_backups.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""In this class, we manage client database relations.
+
+This class creates a user and database for each application relation
+and expose needed information for client connection via fields in
+external relation.
+"""
+import logging
+
+from charms.mongodb.v0.helpers import generate_password
+from charms.mongodb.v0.mongodb import MongoDBConfiguration, MongoDBConnection
+from charms.operator_libs_linux.v1 import snap
+from ops.framework import Object
+from ops.model import ActiveStatus, BlockedStatus
+from pymongo.errors import PyMongoError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "18c461132b824ace91af0d7abe85f40e"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+PBM_S3_CONFIGS = [
+    ("storage.s3.region", "s3-storage-region"),
+    ("storage.s3.bucket", "s3-storage-bucket"),
+    ("storage.s3.prefix", "s3-storage-prefix"),
+    ("storage.s3.credentials.access-key-id", "s3-access-key-id"),
+    ("storage.s3.credentials.secret-access-key", "s3-secret-access-key"),
+    ("storage.s3.serverSideEncryption.kmsKeyID", "s3-kms-key-id"),
+]
+
+PBM_PRIVILEGES = {"resource": {"anyResource": True}, "actions": ["anyAction"]}
+
+
+class MongoDBBackups(Object):
+    """In this class, we manage client database relations."""
+
+    def __init__(self, charm, substrate="k8s"):
+        """Manager of MongoDB client relations."""
+        super().__init__(charm, "client-relations")
+        self.charm = charm
+        self.substrate = substrate
+        self.framework.observe(self.charm.on.config_changed, self._on_pbm_config_changed)
+
+    def _on_pbm_config_changed(self, event) -> None:
+        """Handles PBM configurations."""
+        # handling PBM configurations requires that the pbm snap is installed.
+        if "db_initialised" not in self.charm.app_peer_data:
+            logger.debug("Cannot set PBM configurations, MongoDB has not yet started.")
+            event.defer()
+            return
+
+        snap_cache = snap.SnapCache()
+        pbm_snap = snap_cache["percona-backup-mongodb"]
+
+        if not pbm_snap.present:
+            logger.debug("Cannot set PBM configurations, PBM snap is not yet installed.")
+            event.defer()
+            return
+
+        try:
+            self.create_pbm_user()
+        except PyMongoError as e:
+            self.charm.unit.status = BlockedStatus("config change failed")
+            logger.error("Deferring config_changed since: error=%r", e)
+            event.defer()
```

Comment:
If it defers here, can the charm create the user in the next call to this event?

If so, probably we should change the `BlockedStatus` to some other status as blocked status needs [human/manual intervention](https://juju.is/docs/olm/status-values#heading--workload--charm-status).

I thought on `WaitingStatus` (which I admit is little weird for an error that is being). A better approach could be to set the blocked status only after some retries and then stop deferring or to not defer in this case. Or even separate the errors that could arise in the ones that should block the charm and the ones that can be automatically solved after a defer.